bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java

```diff
@@ -0,0 +1,460 @@
+            newRequest.setHeader("Authorization", "Bearer " + tibberConfig.getToken());
+            newRequest.setSubProtocols("graphql-subscriptions");
+
+            try {
+                client.start();
+                logger.debug("Starting Websocket connection");
+            } catch (Exception e) {
+                logger.debug("Websocket Start Exception: {}", e.getMessage());
+            }
+            try {
+                client.connect(socket, new URI(SUBSCRIPTION_URL), newRequest);
+                logger.debug("Connecting Websocket connection");
+            } catch (IOException e) {
+                logger.debug("Websocket Connect Exception: {}", e.getMessage());
+            } catch (URISyntaxException e) {
+                logger.debug("Websocket URI Exception: {}", e.getMessage());
+            }
+        }
+    }
+
+    public void close() {
+        Session session = this.session;
+        if (session != null) {
+            String disconnect = "{\"type\":\"connection_terminate\",\"payload\":null}";
+            try {
+                socket.sendMessage(disconnect);
+            } catch (IOException e) {
+                logger.debug("Websocket Close Exception: {}", e.getMessage());
+            }
+            session.close();
+            this.session = null;
+        }
+    }
+
+    public boolean isConnected() {
+        Session session = this.session;
+        return session != null && session.isOpen();
+    }
+
+    @WebSocket
+    @NonNullByDefault
+    public class TibberWebSocketListener {
+
+        @OnWebSocketConnect
+        public void onConnect(Session wssession) {
+            TibberHandler.this.session = wssession;
+            logger.debug("Connected to Server");
+
+            String connection = "{\"type\":\"connection_init\", \"payload\":\"token=" + tibberConfig.getToken() + "\"}";
+            try {
+                sendMessage(connection);
+            } catch (IOException e) {
+                logger.warn("Send Message Exception: {}", e.getMessage());
+            }
+        }
+
+        @OnWebSocketClose
+        public void onClose(int statusCode, String reason) {
+            logger.warn("Closing a WebSocket due to {}", reason);
+            TibberHandler.this.session = null;
+            TibberHandler.this.client = null;
+        }
+
+        @OnWebSocketError
+        public void onWebSocketError(Throwable e) {
+            logger.debug("Error during websocket communication: {}", e.getMessage());
+            onClose(0, e.getMessage());
+        }
+
+        @OnWebSocketMessage
+        public void onMessage(String message) {
+            if (message.contains("connection_ack")) {
+                startSubscription();
+            } else if (message.contains("error")) {
+                close();
+                logger.debug("WebSocket Connection closed due to Connection error: {}", message);
+            } else if (!message.contains("error") && message.contains("liveMeasurement")) {
```

Comment:
If it got this far you already know it doesn't contain "error"
```suggestion
            } else if (message.contains("liveMeasurement")) {
```